src/commands/mod.rs

```diff
@@ -2005,14 +2008,71 @@ fn cmd_new(ui: &mut Ui, command: &CommandHelper, args: &NewArgs) -> Result<(), C
         "expected a non-empty list from clap"
     );
     let commits = resolve_base_revs(&workspace_command, &args.revisions)?;
-    let parent_ids = commits.iter().map(|c| c.id().clone()).collect();
+    let parent_ids = commits.iter().map(|c| c.id().clone()).collect_vec();
     let mut tx = workspace_command.start_transaction("new empty commit");
-    let merged_tree = merge_commit_trees(tx.base_repo().as_repo_ref(), &commits);
-    let new_commit = tx
-        .mut_repo()
-        .new_commit(command.settings(), parent_ids, merged_tree.id().clone())
-        .set_description(&args.message)
-        .write()?;
+    let mut num_rebased = 0;
+    let new_commit;
+    if args.insert_before {
+        // Instead of having the new commit as a child of the changes given on the
+        // command line, add it between the changes' parents and the changes.
+        // The parent of the new commit will be the parents of the target commits
+        // which are not descendants of other target commits.
+        let child_ids = parent_ids;
+        let children = RevsetExpression::commits(child_ids.clone());
+        let new_parents = children.parents().minus(&children.descendants());
+        let new_parents_commits = tx
+            .base_workspace_helper()
+            .evaluate_revset(&new_parents)?
+            .iter()
+            .commits(tx.base_repo().store())
+            .collect::<Result<Vec<_>, _>>()?;
+        if new_parents_commits.len() > 1
+            && new_parents_commits.contains(&tx.base_repo().store().root_commit())
+        {
+            return Err(CommandError::UserError {
+                message: "Cannot merge with root revision".to_owned(),
+                hint: None,
+            });
+        }
```

Comment:
Is this check needed? It seems to me that it should be fine to insert a new commit on top of the root commit.

On the other hand, I think we should disallow `jj new --insert-before root`. I think `new_parents_commits.is_empty()` will catch that.

Reply:
The check is still required, even though now it is to remove the root commit from the set (to allow more insertion cases involving root as you wrote, even though it won't be on top of the root commit). See the `test_new_insert_before_no_root_merge` which ensures that root is removed from the set.